### PR TITLE
lib/nolibc,lib/isrlib: Fix strncpy

### PR DESCRIPTION
**Pull request by haesbaert**
strncpy will produce different results for an empty string depending on the preceding value of dst:

char buf[4];

bzero(buf, sizeof(buf));
buf[0] = 0;
buf[1] = 9;
strncpy(&buf[1], "", 3);

"clen" will be 0 since strnlen("") is 0, which will then look into dst[-1] (buf[0]) to decide if it terminates dst or not.

Instead of working around, replace strncpy for a conformant version, taken from OpenBSD libkern, as a bonus we zero out the remaining buffer.

Easier to test by renaming strncpy to something else otherwise the compiler will inline it most of the small cases.

### Prerequisite checklist

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.

### CODE REVIEW

File: lib/nolibc/string.c
Comment (RaduNichita) [inline lib/nolibc/string.c:159]: I have a suggestion here: could we move the zeroing part of the destination after copying the effective data?

Something like:

```c
 while (--len != 0 && (*d++ = *s++) != 0));
 /* NUL pad the remaining n-1 bytes */
 while (--len != 0)
	*d++ = 0;
```

Would it affect the performance in some way?

**Review by RaduNichita — CHANGES_REQUESTED** [review]
Nice work, @haesbaert! I have provided a suggestion regarding the code. 

Furthermore, do you know if the openBSD license is compatible with BSD-3-Clause?  Also, you should add a link to the openBSD implementation of the functions in your commit message.

File: lib/nolibc/string.c
Comment (haesbaert) [reply lib/nolibc/string.c:159]: I don't think it would make any difference, it's worth noting that this is a somewhat inneficcient strncpy, as it copies things in bytes instead of words, an efficient strncpy is a bit of a monstruosity (if you check the ones included in libc's, they're scary!) and don't think it belongs in nolibc

**Comment by haesbaert** [comment]
> Nice work, @haesbaert! I have provided a suggestion regarding the code.
> 
> Furthermore, do you know if the openBSD license is compatible with BSD-3-Clause? Also, you should add a link to the openBSD implementation of the functions in your commit message.

It's the same: https://github.com/openbsd/src/blob/master/sys/lib/libkern/strncpy.c, I'll change the commit message to include it

File: lib/nolibc/string.c
Comment (RaduNichita) [reply lib/nolibc/string.c:159]: I would go for the shortest approach, since it's more easier to follow, in my opinion. @razvand  do you have any opinion regarding this?

Comment (razvand) [reply lib/nolibc/string.c:159]: @RaduNichita , @haesbaert , I don't have a strong opinion. @haesbaert , you should decide what you think is best and that would be it.

File: lib/nolibc/string.c
Comment (RaduNichita) [inline lib/nolibc/string.c:169]: Why do you need brackets here?

**Review by RaduNichita — COMMENTED** [review]
Tested and it has the expected behavior.

I left one question more out of curiosity, I will approve after that

File: lib/nolibc/string.c
Comment (haesbaert) [reply lib/nolibc/string.c:159]: Heya, sorry for the delay I was on holidays.
I'd keep as it is, since it's tested and I don't see an advantage in changing it.

File: lib/nolibc/string.c
Comment (haesbaert) [reply lib/nolibc/string.c:169]: oh the parenthesis are not needed, it's just common in BSD KNF, feel free to remove if wanted.
It's somewhat common in the unikraft codebase though:
```
sam:unikraft: grep -r return\ \( . |wc -l
388
````

**Review by Starnox — APPROVED** [review]
Nice catch! Works for me properly!

Reviewed-by: Eduard-Florin Mihailescu [mihailescu.eduard@gmail.com](mailto:mihailescu.eduard@gmail.com)

File: lib/nolibc/string.c
Comment (skuenzer) [reply lib/nolibc/string.c:169]: Okay, we'll remove the brackets while upstreaming.

**Review by skuenzer — APPROVED** [review]
Approved-by: Simon Kuenzer <simon@unikraft.io>

**Pull request merged**
